devdoc/Metrics.rst

```diff
@@ -0,0 +1,282 @@
+the design of framework components.
+
+Definitions
+===============
+Some definitions are useful.
+
+Metric
+  A measurement defined by a component that represent an internal state that is useful in a system reliability engineering function. In the context
+  of the framework, a metric is an object representing the above.
+
+Metric Value
+  The current value of a metric. A corollary is a metric measurement which represents a view of a metric value that may be based on state.
+
+Collection
+  The process of choosing the relevant metrics and retrieving their values.
+
+Trigger
+  What causes the collection of metric values.
+
+Metric Report
+  Formatted metric values ready for ingestion by a collection system.
+
+
+
+*************
+Use Scenarios
+*************
+This section describes how components expect to use the framework. It is not a complete list of all requirements
+but rather a sample.
+
+
+Roxie
+=====
+Roxie desires to keep a count of many different internal values. Some examples are
+
+* Disk type operations such as seeks and reads
+* Execution totals
+
+  Need to track items such as total numbers of items such as success and failures
+  as well as breaking some counts into individual reasons. For example, failures
+  may need be categorized such as as
+
+  * Busy
+  * Timeout
+  * Bad input
+
+  Or even by priority (high, low, sla)
+
+* Current operational levels such as the length of internal queues
+* The latency of operations such as queue results, agent responses, and gateway responses
+
+Roxie also has the need to track internal memory usage beyond the pod/system level capabilities. Tracking the state of its large fixed memory pool is necessary.
+
+The Roxie buddy system also must track how often and who is completing requests. The "I Beat You To It" set of metrics must be collected and exposed in order to detect pending node failure. While specific action on these counts is not known up front, it appears clear that these values are useful and should be collected.
+
+There does not appear to be a need for creating and destroying metrics dynamically. The set of metrics is most likely to be created at startup and remain active through the life of the Roxie. If, however, stats collection seeps into the metrics framework, dynamic creation and destruction of stats metrics is a likely requirement.
+
+
+ESP
+===
+
+There are some interesting decisions with respect to ESP and collection of metrics. Different applications within ESP present different use cases for collection. Ownership of a given task drives some of these use cases. Take workunit queues. If ownership of the task, with respect to metrics, is WsWorkunits, then uses cases are centric to that component. However, if agents listening on the queue are to report metrics, then a different set of use cases emerge. It is clear that additional work is needed to generate clear ownership of metrics gathered by ESP and/or the tasks it performs.
```

Comment:
My impression of this section is that we don't have a clear idea of what metrics esp needs to collect.  That's fine, but it is worth moving the activeTransactions first, and then note the rest is for future investigation/discussion.

Reply:
Agreed.